BennyTordrupVisma/Week2/Ass3/script.js

```diff
@@ -0,0 +1,340 @@
+	getInitialState: function() {
+		return {
+			//lists: []
+			lists: [{
+				name: "List 1", 
+				messages: [
+					{
+						id: 1,
+						text: "Test 1-1"
+					},
+					{
+						id: 2,
+						text: "Test 1-2",
+						isArchived: true
+					},
+					{
+						id: 3,
+						text: "Test 1-3"
+					},
+					{
+						id: 4,
+						text: "Test 1-4",
+						isArchived: true
+					},
+					{
+						id: 5,
+						text: "Test 1-5"
+					}] 
+			},
+			{
+				name: "List 2", 
+				messages: [
+					{
+						id: 1,
+						text: "Test 2-1"
+					},
+					{
+						id: 2,
+						text: "Test 2-2",
+						isArchived: true
+					},
+					{
+						id: 3,
+						text: "Test 2-3"
+					},
+					{
+						id: 4,
+						text: "Test 2-4",
+						isArchived: true
+					},
+					{
+						id: 5,
+						text: "Test 2-5"
+					}] 
+			}]
+		}
+	},
+	
+	render: function() {
+		var that = this;
+		var listList = this.state.lists.map(function(list, index) {
+			return <List key={index} 
+						list={list} 
+						onArchiveMessage={that.archiveMessage} 
+						onDeleteMessage={that.deleteMessage}
+						onUnarchiveMessage={that.unarchiveMessage}
+						onMoveMessage={that.moveMessage} />
+		});
+		
+		return 	<div>
+					<InputField handleCommit={this.commitMessage} lists={this.state.lists}/>
+					<div>
+						<h3>Lists</h3>
+						<div>
+							{listList}
+						</div>
+					</div>
+				</div>
+	},
+
+	moveMessage: function(message, oldList, newListName) {
+		console.log("Moving message from " + oldList.name + " to " + newListName)
+		var sourceList = this.state.lists.find(l => l.name == oldList.name);
+		if (sourceList == null)
+			return;
+		
+		var destinationList = this.state.lists.find(l => l.name == newListName);
+		if (destinationList == null)
+			return;
+		
+		var msgPos = sourceList.messages.findIndex(m => m.id = message.id);
+		if (msgPos === -1)
+			return;
+
+		var msgText = sourceList.messages[msgPos].text;
+
+		this.deleteMessage(message, oldList);
+		
+		this.commitMessage(msgText, newListName);
+	},
+	
+	deleteMessage: function(message, list) {
+		var listToChange = this.state.lists.find(l => l.name == list.name);
+		if (listToChange == null)
+			return;
+
+		var msgPos = listToChange.messages.findIndex(m => m.id == message.id);
+		if (msgPos === -1)
+			return;
+
+		var tailMessages = listToChange.messages.slice(msgPos + 1);
+		var newMessages = listToChange.messages.slice(0, msgPos).concat(tailMessages);
+		listToChange.messages = newMessages;
+		
+		this.setState({
+			lists: this.state.lists.map(function(list) {
+				return list.name == listToChange.name ? listToChange : list;
+			})
+		});
+	},
+	
+	archiveOrUnarchiveMessageKernel: function(message, list, doArchive) {
+		var listToChange = this.state.lists.find(l => l.name == list.name);
+		if (listToChange == null)
+			return;
+		
+		var msgToArchive = listToChange.messages.find(m => m.id == message.id);
+		if (msgToArchive == null)
+			return;
+		
+		msgToArchive.isArchived = doArchive;
```

Comment:
`msgToArchive` still points to the state object, so here you are mutating state directly. You can use `Object.assign({}, msgToArchive)` for cloning an object. Please see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign